[BEAM-3702] adding fromJvm to create pipelineoptions from the system properties

rmannibucau opened the pull request:
Adds a factory method in `PipelineOptionsFactory` to instantiate the `PipelineOptions` from the "JVM" which means using the environment and system properties (in this order to support overriding).

It also adds a factory from plain properties to let users read it the way they want.
To be flexible enough, and as discussed on the list, the factory can take a prefix to filter the properties
and if not is provided the default is "beam.".

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand:
   - [ ] What the pull request does
   - [ ] Why it does it
   - [ ] How it does it
   - [ ] Why this approach
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

kennknowles commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -1761,4 +1763,67 @@ private static void checkEmptyStringAllowed(Class<?> type, JavaType genericType,
+   * @return a pipeline options instance based on the specified properties.
+   */
+  public static PipelineOptions fromProperties(final Properties properties) {
+    if (properties == null) {
```
Properties cannot be null. Our codebase has the policy that nothing is nullable unless explicitly marked. And here there is no benefit to doing that.

rmannibucau replied:
Hmm, assuming this will be used under the hood and beam can't assume the properties instantiation/lookup I'd like to accept null here. Should I just add @Nullable?

kennknowles replied:
I don't understand your comment. I don't think it makes sense for any of these to accept `null`.

rmannibucau replied:
It is about making like easy for end users.

Typical code will be:

    Properties p = loadCustomProps(); // can be null, no way beam can help here
    return fromProperties(p);

To avoid a null check in all user apps/lib I d prefer to put it in a single place, ie beam.

kennknowles commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -1761,4 +1763,67 @@ private static void checkEmptyStringAllowed(Class<?> type, JavaType genericType,
+   * filtered with the specified prefix.
+   */
+  public static PipelineOptions fromProperties(final String prefix, final Properties properties) {
+    if (prefix == null) {
```
ditto: cannot be null

kennknowles commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -1761,4 +1763,67 @@ private static void checkEmptyStringAllowed(Class<?> type, JavaType genericType,
+    if (prefix == null) {
+      return fromProperties(properties);
+    }
+    if (properties == null) {
```
cannot be null

kennknowles commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -1761,4 +1763,67 @@ private static void checkEmptyStringAllowed(Class<?> type, JavaType genericType,
+      .stream()
+      .filter(k -> k.startsWith(prefix))
+      .collect(
+        Properties::new,
```
Is there any reason to collect this into a `Properties` object instead of just a `Map` or even just `List` of entries?

rmannibucau replied:
overall idea is to have a single base signature. Choice is between Properties, Map<String, String>, List<Pair<String, String>> (or entry). I picked Properties because it is easy for system properties and to read from any source (properties.load) without any external dependency. Since it is a volatile data used in a single thread there is no concurrency issue so I choose to convert the only map manipulated by this code to a properties instance to conform to the other signatures.

kennknowles replied:
I guess my thought is that you should convert to a more basic type ASAP since properties are not relevant except for as an input format (just like command line is not relevant except as an input format). In this case, Properties is also basically just a Map anyhow so it doesn't bother me so much.

The fact that it is Map<Object, Object> may either be a bug or a feature; I'm not sure yet.

rmannibucau replied:
Properties is an niput format yes, it is aligned with fromArgs and factories here which convert an inputformat to an options instance. Guess we can split it in N factories bit maybe overkill here. The object typing is inherited from a pre generics time but it should really be string/string and this is xhy i use the string keying which already filters it properly.

Map of strings can be used but is way less natural in end user code.

Same as previous comment: this is for users, not for beam dev so trying to embrace their future usage rather our own satisfaction.

kennknowles commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -1761,4 +1763,67 @@ private static void checkEmptyStringAllowed(Class<?> type, JavaType genericType,
+   * @return a pipeline options instance based on current environment and overriden with
+   * system properties key/values and filtered with a custom prefix.
+   */
+  public static PipelineOptions fromJvm(final String prefix) {
```
`fromSystemProperties`

rmannibucau replied:
notr really it is the env overriden with the system properties, so it is really fromJvm. Strictly speaking it could be fromEnvironmentAndSystemProperties but it is way too long to be sane ;).

kennknowles replied:
Don't override it from the env :-)

rmannibucau replied:
Sorry to ask another time Kenn but I dont get the rational, this is a very common strategy all config projects use, why would it break it?

kennknowles commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -1761,4 +1763,67 @@ private static void checkEmptyStringAllowed(Class<?> type, JavaType genericType,
+    }
+    final String[] args = properties.stringPropertyNames()
+      .stream()
+      .map(k -> String.format("--%s=%s", k, properties.getProperty(k)))
```
Don't use the command line parsing; this isn't a command line. Directly create the PipelineOptions.

rmannibucau replied:
I can miss something but only args are parsed to this just builds args to ensure we keep a single instantiation mecanism. I don't want to go through a JSON model here but really reuse the plain args parsing to enforce consistency accross the usages.

kennknowles replied:
There should be a single shared mechanism between args and properties, yes. It should accept a map or some structured value - I haven't dug in this code for a while. But a fake cmdline is not how to instantiate an object.

rmannibucau replied:
What is the issue? I have to admit I kind of like it enforces CLI support is not broken and cant be. The alternative makes it very tempting to not respect it.

kennknowles commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -1761,4 +1763,67 @@ private static void checkEmptyStringAllowed(Class<?> type, JavaType genericType,
+   * @return a pipeline options instance based on current environemnt and system properties
+   * filtered based on <code>beam.</code> prefix.
+   */
+  public static PipelineOptions fromJvm() {
```
Don't add this. Each use needs to choose its own prefix. The test pipeline can read from `beam.TestPipelineOptions` and that prefix should be owned over there, not here. Any user should always have to specify a prefix on the system properties.

rmannibucau replied:
I don't think so, it is very important to propose a default prefix (beam.) to enable frameworks to have a default and uniform implicit usage. This doesn't mean we should encapsulate the prefix but that we should have an imlpicit prefix when the integration framework don't care much. If not done each framework will have its own prefix which wouldn't help much users in several cases.

wdyt?

kennknowles replied:
PipelineOptions are an explicit per-job property. Only a program that runs exactly one job could use this, and that program can just specify a prefix. Again, this is different to the discussion on dev. If you want to set up global default pipeline options via properties that is a separate thing that just supporting parsing them. Bring back to the dev list.

rmannibucau replied:
Not really, test flavor proves it is not about only one job. Also it concerns only the submit so even with the one job constraint it concers a lot of users so is worth it IMHO. Will add it in the mail, thanks.

kennknowles commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -1761,4 +1763,67 @@ private static void checkEmptyStringAllowed(Class<?> type, JavaType genericType,
+              Properties::new,
+              (p, e) -> p.setProperty(e.getKey(), e.getKey()),
+              Properties::putAll);
+    properties.putAll(System.getProperties());
```
Don't blend JVM properties with environment variables. For this feature, just use system properties. The way to interact with the environment is to put the whole JSON into one variable.

rmannibucau replied:
This way is more convenient for a lot of people. I saw nobody to be able to use properly and reliably the JSON flavor cause depending where you set it the escaping is different (console, IDE, shell scripts, ...). Using the env enables ops to use that *naturally*, using the system props enables the dev to use it *naturally*. This is where it is coming from.

kennknowles replied:
This particular bit of the PR is not agreed on. I don't like it - it seems like a real abuse of env vars. There could be a zillion pipeline options and it is very weird to have a cluster of env vars parsed together like this. I could be OK with it but I want the dev list to discuss.

If you do want to also support this, do it in a separate PR, separate method, and get some feedback on the dev list about that.

rmannibucau replied:
I ll ping today the list. If a discusdion is needed ill split it, ok?

rmannibucau commented:
removed null checks and env usage, added Map<String, String> support - I'd like to keep Properties bridges since it is how you will get it in practise, and added a fromMap. Hope it fixes issues you point out. We'll see on the list for other points.

rmannibucau commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -185,19 +190,21 @@ public Builder withValidation() {
   public static class Builder {
     private final String defaultAppName;
     private final String[] args;
+    private final Map<String, String> options;
```
Hmm, it is not about converting a command line to a map but taking a map as input to build the map<String, Object> (the generics are important here) which are the backing instances of the proxy.

Both are merged in a single source which is then converted in instances used by the proxy.
The internal format is a multimap but I don't want to put that in an API so the Map<String, String> is at the exact same level than the String[] IMHO, no?

kennknowles replied:
Command line is a `Map<String, String>` so that's the common denominator. I know properties support something like `Map<String, Object>` but for properties files and `-D` it is again a `Map<String, String>`. The conversion from `Map<String, String>` to `Map<String, Object>` is what `parseObjects` does.

rmannibucau replied:
-D is not Map<String, String> but a multimap since you can pass N times the same option so we dont have a common structure except Map<String, Object>. Will try to move to that as an internal.

kennknowles commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -189,21 +202,19 @@ public Builder withValidation() {
 
-    private Builder(String[] args, Map<String, String> options, boolean validation,
-        boolean strictParsing, boolean isCli) {
+    private Builder(Collection<Function<Builder, Multimap<String, String>>> options,
```
`Properties` isn't a multimap, actually. Entries overwrite each other when they are specified more than once.

rmannibucau replied:
Hmm, thought https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java#L1546 concatenate it when parsing the args and values are differents (means --foo=a --foo=a would loose a value compared to --foo=a --foo=b).

kennknowles replied:
Ah, makes sense. So cmdline is multimap while Properties are limited to just a map. SGTM

kennknowles replied:
TBH it is not a good cmdline interface to do this

rmannibucau replied:
well tempted to say +-0, I saw so much times the debate = vs space to delimit the key/value and single option with N values vs N times the options that I just pick one now and move forward ;).

lukecwik commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -1577,7 +1589,7 @@ public int compare(Method o1, Method o2) {
    */
   private static <T extends PipelineOptions> Map<String, Object> parseObjects(
-      Class<T> klass, ListMultimap<String, String> options, boolean strictParsing) {
+      Class<T> klass, Multimap<String, String> options, boolean strictParsing) {
```
Its important that this is a listmultimap and not a setmultimap as the command line "--foo=a --foo=a" should be equivalent to "--foo=a,a". Also, the parameters on the command line should not be re-ordered.

kennknowles replied:
If that's the spec, then `List<Pair<String, List<String>>>` makes the spec in the type, which is a bit better.

rmannibucau replied:
@lukecwik agree, multimap is just more convenient and since it is an internal it is fine. If you check org.apache.beam.sdk.options.PipelineOptionsFactory.Builder#as I really use a list flavor of the multimap to guarantee it and parseCommandLine does too. The map case doesnt care. So we are all good.
@kennknowles for this internal structure `List<Pair<String, List<String>>>` is harder to use cause you miss the put for the usage which is done of the structure, probably better to avoid useless loops in the manipulation, no?

lukecwik replied:
Multimap is to vague on the contract, please stick with ListMultimap since it clearly specifies it.

This also documented in the Javadoc for Multimap: "Instead of using the Multimap interface directly, prefer the subinterfaces ListMultimap and SetMultimap. These take their names from the fact that the collections they return from get behave like (and, of course, implement) List and Set, respectively."

iemejia commented:
Hi, just curious if it makes sense to mark these new methods as @Experimental at least until we are sure the names/contracts are right and they won't change. (Note, you can just ignore me if you are convinced it won't, just telling because these are new sdks/core things so contract is stricter).

rmannibucau commented:
@iemejia no strong opinion on that, planning to move to listmultimap later today to finish it so can add @exp at the same time

kennknowles commented:
Yes, good point @iemejia. These should all be `@Experimental`.

rmannibucau commented:
updated code for listmultimap and @Experimental points

iemejia commented:
LGTM overall and about my comment on Experimental. @rmannibucau please rebase because there are merge issues.

kennknowles commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -171,6 +183,13 @@ public static Builder fromArgs(String... args) {
+  /**
+   * @return a builder instance enabling you to build pipeline options.
+   */
+  public static Builder builder() {
```
Unused and it looks like the other initialization methods handle it.

rmannibucau replied:
Yes and no. Idea is to provide a clean builder entry point. Others are just shortcuts but way less relevant now you have multiple (2) entry points you can combine in the order you want to handle overrides accordingly your need. Alternative is to drop the fromArgs and provide a argsToMap utility but it breaks the compatibility.

kennknowles replied:
In that case, just a little nit that they should be `addMap` and `addArgs`. This way it is clear that they keep the existing options and add new ones.

rmannibucau replied:
Hmm, do we break the API then? (on a personnal note/feeling: a builder with withers sounds as good)

rmannibucau commented:
rebased

lukecwik commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -245,7 +263,19 @@ private Builder(String[] args, boolean validation,
+
+    /**
+     * A key/value set of options.
+     * @param configuration the options to convert to PipelineOptions.
```
Please fill out a comment stating how the parsing works similar to fromArgs()

lukecwik commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -245,7 +263,19 @@ private Builder(String[] args, boolean validation,
     public Builder fromArgs(String... args) {
       checkNotNull(args, "Arguments should not be null.");
-      return new Builder(args, validation, strictParsing, true);
+      options.add(current -> parseCommandLine(args, current.strictParsing));
```
It is important to point out that this is a subtle change in behavior. Before fromArgs replaced the existing set of args and now we are accumulating them. Please update the javadoc accordingly.

lukecwik commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -1182,7 +1194,7 @@ private static void validateInheritedInterfacesExtendPipelineOptions(Class<?> kl
 
   private static class MultipleDefinitions {
     private Method method;
-    private SortedSet<Method> collidingMethods;
```
Its important that this remains being sorted for logging reasons.

lukecwik commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -1792,4 +1804,82 @@ private synchronized void register(Class<? extends PipelineOptions> iface) {
+   * @return a pipeline options instance based on the specified configuration.
+   */
+  @Experimental
+  public static PipelineOptions fromMap(final String prefix, final Map<String, String> options) {
```
Why is this important to have part of the API?

It seems like we should ask users to filter their map keys themselves.

lukecwik commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -1792,4 +1804,82 @@ private synchronized void register(Class<? extends PipelineOptions> iface) {
+  }
+
+  /**
+   * @param options the options to convert to a pipeline options.
```
Please improve the javadoc to be similar to `fromArgs` explaining the parsing.

lukecwik commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -1792,4 +1804,82 @@ private synchronized void register(Class<? extends PipelineOptions> iface) {
+  }
+
+  /**
+   * @param properties the properties to use to create the pipeline options instance.
```
Please improve the javadoc to be similar to `fromArgs` explaining the parsing.

It's also important to mention that you filter out non string key/value pairs since many people may assume that might call toString on everything.

lukecwik commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -1792,4 +1804,82 @@ private synchronized void register(Class<? extends PipelineOptions> iface) {
+
+  /**
+   * @param prefix the prefix filter applied on system properties keys.
+   * @return a pipeline options instance based on system properties key/values
```
Please improve/explain what is the contract for this method.

lukecwik commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -1792,4 +1804,82 @@ private synchronized void register(Class<? extends PipelineOptions> iface) {
+
+  /**
+   * @param prefix the prefix filter applied on system properties keys.
+   * @param properties the properties to use to create the pipeline options instance.
```
Please improve the javadoc to be similar to `fromArgs` explaining the parsing.

It's also important to mention that you filter out non string key/value pairs since many people may assume that might call toString on everything. Please also mention how prefix is used.

lukecwik commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -1792,4 +1804,82 @@ private synchronized void register(Class<? extends PipelineOptions> iface) {
+
+  /**
+   * @return a pipeline options instance based on system properties
+   * filtered based on <code>beam.</code> prefix.
```
Please improve the contract of this method.

lukecwik commented on sdks/java/core/src/test/java/org/apache/beam/sdk/options/PipelineOptionsFactoryTest.java:
```diff
@@ -1797,6 +1799,44 @@ public void serialize(JacksonIncompatible jacksonIncompatible, JsonGenerator jso
+    assertEquals(
+      "testAppName",
+      PipelineOptionsFactory.fromProperties("prefix.", new Properties() {{
+        put("prefix.appName", "testAppName");
```
Add a value which shouldn't be matched and test that it isn't set.

lukecwik commented on sdks/java/core/src/test/java/org/apache/beam/sdk/options/PipelineOptionsFactoryTest.java:
```diff
@@ -1797,6 +1799,44 @@ public void serialize(JacksonIncompatible jacksonIncompatible, JsonGenerator jso
+  }
+
+  @Test
+  public void testPipelineOptionsFactoryFromPropertiesWithLongPrefix() {
```
This test doesn't add value.

lukecwik commented on sdks/java/core/src/test/java/org/apache/beam/sdk/options/PipelineOptionsFactoryTest.java:
```diff
@@ -1797,6 +1799,44 @@ public void serialize(JacksonIncompatible jacksonIncompatible, JsonGenerator jso
+
+  @Test(expected = IllegalArgumentException.class)
+  public void testPipelineOptionsFactoryFromNullProperties() {
+    assertNotNull(PipelineOptionsFactory.fromProperties(null));
```
Drop the assertNotNull, it doesn't do anything

lukecwik commented on sdks/java/core/src/test/java/org/apache/beam/sdk/options/PipelineOptionsFactoryTest.java:
```diff
@@ -1797,6 +1799,44 @@ public void serialize(JacksonIncompatible jacksonIncompatible, JsonGenerator jso
+
+  @Test(expected = IllegalArgumentException.class)
+  public void testPipelineOptionsFactoryFromPropertiesAndNullPrefix() {
+    assertEquals(
```
drop the assertEquals as it does nothing

lukecwik commented on sdks/java/core/src/test/java/org/apache/beam/sdk/options/PipelineOptionsFactoryTest.java:
```diff
@@ -1797,6 +1799,44 @@ public void serialize(JacksonIncompatible jacksonIncompatible, JsonGenerator jso
+      PipelineOptionsFactory.fromProperties(null, new Properties() {{
+        put("appName", "testAppName");
+      }}).as(ApplicationNameOptions.class).getAppName());
+  }
```
nit: whitespace

lukecwik commented on sdks/java/core/src/test/java/org/apache/beam/sdk/options/PipelineOptionsFactoryTest.java:
```diff
@@ -1797,6 +1799,44 @@ public void serialize(JacksonIncompatible jacksonIncompatible, JsonGenerator jso
+      "testAppName",
+      PipelineOptionsFactory.fromProperties(null, new Properties() {{
+        put("appName", "testAppName");
+      }}).as(ApplicationNameOptions.class).getAppName());
```
Please add tests for system properties

lukecwik commented on sdks/java/core/src/test/java/org/apache/beam/sdk/options/PipelineOptionsFactoryTest.java:
```diff
@@ -1797,6 +1799,44 @@ public void serialize(JacksonIncompatible jacksonIncompatible, JsonGenerator jso
 
+  @Test
+  public void testPipelineOptionsFactoryFromProperties() {
+    assertEquals("testAppName", PipelineOptionsFactory.fromProperties(new Properties() {{
```
Add a string key with non string value to the properties and also a non string key and string value and validate that they aren't used.

rmannibucau commented:
addressed the changes and merged with master. hope it is fine this time :)

lukecwik commented:
Please fix up the javadoc errors:
2018-03-18T19:18:01.876 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:3.0.0-M1:jar (attach-javadocs) on project beam-sdks-java-core: MavenReportException: Error while generating Javadoc:
2018-03-18T19:18:01.876 [ERROR] Exit code: 1 - /home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_MavenInstall@2/src/sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:1924: error: unexpected end tag: </p>
2018-03-18T19:18:01.876 [ERROR] * key/values.</p>
2018-03-18T19:18:01.876 [ERROR] ^
2018-03-18T19:18:01.876 [ERROR] /home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_MavenInstall@2/src/sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:276: error: unexpected end tag: </p>
2018-03-18T19:18:01.876 [ERROR] * as a prefix in this case.</p>
2018-03-18T19:18:01.876 [ERROR] ^

rmannibucau commented:
fixed

lukecwik commented:
Please fix the javadoc errors:
```
[ERROR] /home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_MavenInstall/src/sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:273: Redundant <p> tag. [JavadocParagraph]
Audit done.
2018-03-21T21:43:26.135 [INFO] There is 1 error reported by Checkstyle 8.7 with beam/checkstyle.xml ruleset.
2018-03-21T21:43:26.138 [ERROR] src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:[273] (javadoc) JavadocParagraph: Redundant <p> tag.
```
